zephyr/lib/fast-get.c

```diff
@@ -0,0 +1,164 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2024 Intel Corporation. All rights reserved.
+//
+// Author: Jyri Sarha <jyri.sarha@linux.intel.com>
+
+#include <stdbool.h>
+#include <stdint.h>
+#include <errno.h>
+
+#include <sof/lib/fast-get.h>
+#include <rtos/alloc.h>
+#include <rtos/spinlock.h>
+#include <rtos/symbol.h>
+#include <ipc/topology.h>
+
+struct sof_fast_get_entry {
+	const void *dram_ptr;
+	void *sram_ptr;
+	size_t size;
+	unsigned int refcount;
+};
+
+struct sof_fast_get_data {
+	struct k_spinlock lock;
+	size_t num_entries;
+	struct sof_fast_get_entry *entries;
+};
+
+static struct sof_fast_get_data fast_get_data = {
+	.num_entries = 0,
+	.entries = NULL,
+};
+
+LOG_MODULE_REGISTER(fast_get, CONFIG_SOF_LOG_LEVEL);
+
+static int fast_get_realloc(struct sof_fast_get_data *data)
+{
+	struct sof_fast_get_entry *entries;
+
+	if (!data->num_entries) {
+		data->entries = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM,
+					32 * sizeof(*entries));
+		if (!data->entries)
+			return -ENOMEM;
+		data->num_entries = 32;
+		return 0;
+	}
+
+	entries = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM,
+			  2 * data->num_entries * sizeof(*entries));
+	if (!entries)
+		return -ENOMEM;
+
+	memcpy_s(entries, 2 * data->num_entries * sizeof(*entries), data->entries,
+		 data->num_entries * sizeof(*entries));
+	rfree(data->entries);
+	data->entries = entries;
+	data->num_entries = 2 * data->num_entries;
+	return 0;
+}
+
+static struct sof_fast_get_entry *fast_get_find_entry(struct sof_fast_get_data *data,
+						      const void *dram_ptr)
+{
+	int i;
+
+	for (i = 0; i < data->num_entries; i++) {
+		if (data->entries[i].dram_ptr == dram_ptr)
+			return &data->entries[i];
+	}
+
+	for (i = 0; i < data->num_entries; i++) {
+		if (data->entries[i].dram_ptr == NULL)
+			return &data->entries[i];
+	}
+
+	return NULL;
+}
+
+const void *fast_get(const void *dram_ptr, size_t size)
+{
+	struct sof_fast_get_data *data = &fast_get_data;
+	struct sof_fast_get_entry *entry;
+	k_spinlock_key_t key;
+	void *ret;
+
+	key = k_spin_lock(&data->lock);
+	do {
+		entry = fast_get_find_entry(data, dram_ptr);
+		if (!entry) {
+			if (fast_get_realloc(data)) {
+				ret = NULL;
+				goto out;
+			}
+		}
+	} while (!entry);
+
+	if (entry->sram_ptr) {
+		if (entry->size != size || entry->dram_ptr != dram_ptr) {
+			tr_err(fast_get, "size %u != %u or ptr %p != %p mismatch",
+				entry->size, size, entry->dram_ptr, dram_ptr);
+			ret = NULL;
+			goto out;
+		}
+
+		ret = entry->sram_ptr;
+		entry->refcount++;
+		goto out;
+	}
+
+	ret = rmalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, size);
+	if (!ret)
+		goto out;
+	entry->size = size;
+	entry->sram_ptr = ret;
+	memcpy_s(entry->sram_ptr, entry->size, dram_ptr, size);
```

Comment:
Are we sure size is less than entry->size always (as size is not passed to find_entry())?